### PR TITLE
Group Data Provider: RemoveFabric method added.

### PR DESCRIPTION
**Pull request by rcasallas-silabs**
#### Problem
The GroupDataProvider interface is missing a function to remove all entries related to a given fabric.

#### Change overview
* Added RemoveFabric method to GroupDataProvider
* Added unit test for RemoveFabric

#### Testing
Unit tests added with coverage for the feature, and run successfully.

### CODE REVIEW

File: src/credentials/examples/GroupDataProviderExample.cpp [hunk 1]
```diff
@@ -516,13 +531,15 @@ class StaticGroupsProvider : public GroupDataProvider
...
-    CHIP_ERROR SetKeySet(chip::FabricIndex fabric_index, uint16_t key_set_index, KeySet & keys) override
+    CHIP_ERROR SetKeySet(chip::FabricIndex fabric_index, uint16_t key_set_index, const KeySet & keys) override
     {
         VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
 
-        Fabric * fabric   = GetExistingFabric(fabric_index);
+        Fabric * fabric   = GetExistingFabricOrAllocateNew(fabric_index);
         KeysEntry * entry = nullptr;
 
+        VerifyOrReturnError(fabric, CHIP_ERROR_INVALID_FABRIC_ID);
```
Comment (tcarmelveilleux) [inline src/credentials/examples/GroupDataProviderExample.cpp:541]:
```suggestion
        VerifyOrReturnError(fabric != nullptr, CHIP_ERROR_INVALID_FABRIC_ID);
```

File: src/credentials/examples/GroupDataProviderExample.cpp [hunk 2]
```diff
@@ -599,6 +620,23 @@ class StaticGroupsProvider : public GroupDataProvider
...
+        Fabric * fabric = GetExistingFabric(fabric_index);
+        VerifyOrReturnError(fabric, CHIP_ERROR_INVALID_FABRIC_ID);
+        // Remove group states
+        for (size_t i = 0; i < kMaxNumGroupStates; ++i)
+        {
+            if (mGroupStates[i].fabric == fabric)
+            {
+                RemoveGroupState(i);
+            }
+        }
```
Comment (tcarmelveilleux) [inline src/credentials/examples/GroupDataProviderExample.cpp:634]: We must also remove the keysets for the fabric

File: src/credentials/examples/GroupDataProviderExample.cpp [hunk 3]
```diff
@@ -599,6 +620,23 @@ class StaticGroupsProvider : public GroupDataProvider
...
+        // Release fabric entry
```
Comment (tcarmelveilleux) [inline src/credentials/examples/GroupDataProviderExample.cpp:635]:
```suggestion
        // Release other resources associated with the fabric entry, such as mapped
        // endpoints (i.e. through the Groups cluster) and group key sets, both of which
        // live in the `Fabric` data structure of this implementation.
```

File: src/credentials/tests/TestGroupDataProvider.cpp [hunk 4]
```diff
@@ -582,6 +579,246 @@ void TestKeysIterator(nlTestSuite * apSuite, void * apContext)
     groups->Finish();
 }
 
+void TestFabrics(nlTestSuite * apSuite, void * apContext)
```
Comment (tcarmelveilleux) [inline src/credentials/tests/TestGroupDataProvider.cpp:582]: Suggest "TestPerFabricData"

File: src/credentials/tests/TestGroupDataProvider.cpp [hunk 5]
```diff
@@ -582,6 +579,246 @@ void TestKeysIterator(nlTestSuite * apSuite, void * apContext)
...
+    exists = groups->GroupMappingExists(kFabricIndex2, group1c);
+    NL_TEST_ASSERT(apSuite, exists);
+
+    // States
+
+    err = groups->GetGroupState(0, state0b);
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+    NL_TEST_ASSERT(apSuite, state1a.fabric_index == state0b.fabric_index);
+    NL_TEST_ASSERT(apSuite, state1a.group == state0b.group);
+    NL_TEST_ASSERT(apSuite, state1a.key_set_index == state0b.key_set_index);
```
Comment (tcarmelveilleux) [inline src/credentials/tests/TestGroupDataProvider.cpp:779]: It's not easy to understand, from the naming convention here, why those should match after RemoveFabric.

**Review by jmartinez-silabs — APPROVED** [review]

File: src/credentials/tests/TestGroupDataProvider.cpp [hunk 6]
```diff
@@ -582,6 +579,246 @@ void TestKeysIterator(nlTestSuite * apSuite, void * apContext)
...
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->SetKeySet(kFabricIndex1, 606, keys0a);
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+
+    err = groups->GetKeySet(kFabricIndex2, 101, keys_out);
+    NL_TEST_ASSERT(apSuite, CHIP_NO_ERROR == err);
+    NL_TEST_ASSERT(apSuite, keys0a.policy == keys_out.policy);
+    NL_TEST_ASSERT(apSuite, keys0a.num_keys_used == keys_out.num_keys_used);
+    NL_TEST_ASSERT(apSuite, !memcmp(keys0a.epoch_keys, keys_out.epoch_keys, sizeof(keys0a.epoch_keys[0]) * keys0a.num_keys_used));
```
Comment (emargolis) [inline src/credentials/tests/TestGroupDataProvider.cpp:717]:
```suggestion
    NL_TEST_ASSERT(apSuite, 0 == memcmp(keys0a.epoch_keys, keys_out.epoch_keys, sizeof(keys0a.epoch_keys[0]) * keys0a.num_keys_used));
```

**Review by emargolis — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Pull request merged**